doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,326 @@
+ * Pluggable storage backends
+
+## Data Types
+
+### Result
+
+This is the result made available over watchers, `Get()` etc. All backends must return an implementation providing at least this information.
+
+```go
+type Result interface {
+	// Bucket is the bucket the data was loaded from
+	Bucket() string
+	// Key is the key that was retrieved
+	Key() string
+	// Value is the retrieved value
+	Value() string
+	// Created is the time the data was received in the bucket
+	Created() time.Time
+	// Sequence is a unique sequence for this value
+	Sequence() uint64
+	// Delta is distance from the latest value. If history is enabled this is effectively the index of the historical value, 0 for latest, 1 for most recent etc.
+	Delta() uint64
+	// OriginCluster is the cluster where this data originate from, may be empty if sharing was not enabled
+	OriginCluster() string
+	// Operation is the kind of operation this result represents, enum of PUT and DEL
+	Operation() Operation
+}
+```
+
+### Status
+
+This is the status of the KV as a whole
+
+```go
+type Status interface {
+	// Bucket the name of the bucket
+	Bucket() string
+
+	// Values is how many messages are in the bucket, including historical values
+	Values() uint64
+
+	// History returns the configured history kept per key
+	History() int64
+
+	// TTL is how long the bucket keeps values for
+	TTL() time.Duration
+
+	// Cluster returns the name of the cluster holding the read replica of the data
+	Cluster() string
+
+	// Replicas returns how many times data in the bucket is replicated at storage
+	Replicas() (ok int, failed int)
+
+	// Keys returns a list of all keys in the bucket - not possible now
+	Keys() ([]string, error)
+
+	// BackingStore is a backend specific name for the underlying storage - eg. stream name
+	BackingStore() string
+
+	// MirrorStatus is the status of a read replica, error when not accessing a replica (planned)
+	MirrorStatus() (lag int64, active bool, err error)
+}
+```
+
+## RoKV
+
+This is a read-only KV store handle, every backend should implement a language equivalent interface
+
+```go
+// RoKV is a read-only interface to a single key-value store bucket
+type RoKV interface {
+	// Get gets a key from the store
+	Get(key string) (Result, error)
+
+	// History retrieves historic values for a key
+	History(ctx context.Context, key string) ([]Result, error)
+
+	// WatchBucket watches the entire bucket for changes, all keys and values will be traversed including all historic values
+	WatchBucket(ctx context.Context) (Watch, error)
+
+	// Watch a key for updates, the same Result might be delivered more than once
+	Watch(ctx context.Context, key string) (Watch, error)
+
+	// Close releases in-memory resources held by the KV, called automatically if the context used to create it is canceled
+	Close() error
+
+	// JSON dumps the entire KV as k=v values in JSON format
+	JSON(ctx context.Context) ([]byte, error)
```

Comment:
Need to think through this one more ;)

Reply:
unfortunately given how this works, its what we need - its a push consumer and if ctx timeout before we reach end we need to bail out.

anyway, as with `[]byte` whatever makes a specific language happy as long as you make the bucket available - for the CLI this works for me.

Reply:
Yea but this is too simplistic and only works for small data sets. Need to think more about this IMO.

Reply:
Do we need to limit value to string?  Nats allow slice of bytes for Data and I think that’s needed for a kv store as well.

Reply:
@tpihl As mentioned previously, languges can implement this however they like - of course I convert string to byte here because the reference implementation is there to support the CLI.

Language authors can do the right thing for their language.  I am not going to dictate to C/C++/Rust/Java/.Net to use a thing, they know what makes sense for their case.  This is a guide. Langauge implementers need to make their appropriate choices.

I was borrowing many concepts and ideas from etcd for example, their go API has 

```
ctx, cancel := context.WithTimeout(context.Background(), timeout)
resp, err := kvc.Put(ctx, "sample_key", "sample_value")
```